src/CaptureService/CaptureServiceBase.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "CaptureService/CaptureServiceBase.h"
+
+#include <absl/time/time.h>
+
+#include "CaptureService/CommonProducerCaptureEventBuilders.h"
+#include "GrpcProtos/Constants.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Profiling.h"
+
+using orbit_grpc_protos::CaptureOptions;
+using orbit_grpc_protos::ProducerCaptureEvent;
+
+using orbit_producer_event_processor::ClientCaptureEventCollector;
+using orbit_producer_event_processor::ProducerEventProcessor;
+
+namespace orbit_capture_service {
+
+CaptureServiceBase::CaptureServiceBase() {
+  // We want to estimate clock resolution once, not at the beginning of every capture.
+  clock_resolution_ns_ = orbit_base::EstimateAndLogClockResolution();
+}
+
+void CaptureServiceBase::AddCaptureStartStopListener(CaptureStartStopListener* listener) {
+  bool new_insertion = capture_start_stop_listeners_.insert(listener).second;
+  ORBIT_CHECK(new_insertion);
+}
+
+void CaptureServiceBase::RemoveCaptureStartStopListener(CaptureStartStopListener* listener) {
+  bool was_removed = capture_start_stop_listeners_.erase(listener) > 0;
+  ORBIT_CHECK(was_removed);
+}
+
+CaptureServiceBase::CaptureInitializationResult CaptureServiceBase::InitializeCapture(
+    ClientCaptureEventCollectorBuilder* client_capture_event_collector_builder) {
+  {
+    absl::MutexLock lock(&capture_mutex_);
+    if (is_capturing_) {
+      return CaptureInitializationResult::kAlreadyInProgress;
+    }
+    is_capturing_ = true;
+  }
+
+  client_capture_event_collector_ =
+      client_capture_event_collector_builder->BuildClientCaptureEventCollector();
+  producer_event_processor_ = ProducerEventProcessor::Create(client_capture_event_collector_.get());
+  return CaptureInitializationResult::kSuccess;
+}
+
+void CaptureServiceBase::TerminateCapture() {
+  producer_event_processor_.reset();
+  client_capture_event_collector_.reset();
+  capture_start_timestamp_ns_ = 0;
+
+  absl::MutexLock lock(&capture_mutex_);
+  is_capturing_ = false;
+}
+
+void CaptureServiceBase::StartEventProcessing(const CaptureOptions& capture_options) {
+  // These are not in precise sync but they do not have to be.
+  absl::Time capture_start_time = absl::Now();
+  capture_start_timestamp_ns_ = orbit_base::CaptureTimestampNs();
+
+  producer_event_processor_->ProcessEvent(
+      orbit_grpc_protos::kRootProducerId,
+      orbit_capture_service::CreateCaptureStartedEvent(capture_options, capture_start_time,
```

Comment:
Mostly a note: I assume this was there before, but it feels strange that the CreateCaptureStartedEvent gets two different capture start timestamps...

Reply:
Yeah, I thought the same. It's wall clock time and steady clock if I remember correctly. Maybe it's worthwhile to rename those to make the distinction more clear. But probably out of scope of this PR.

Reply:
It was there before. I'm not clear about the reason for doing this. Seems that:
* `capture_start_timestamp_ns` is derived by converting a raw timestamp retrieved from `QueryPerformanceCounter()` to nanoseconds, and it is used to set `CaptureStarted::capture_start_timestamp_ns`. This is the timestamp used to place events on Orbit's main capture timeline.
* `capture_start_time` is derived by converting an `absl::Time` from the system clock to unix nanoseconds, and it is used to set `CaptureStarted::capture_start_unix_time_ns`. This is only used in capture log.

Is there any difference for extracting current time with `QueryPerformanceCounte` and `absl::Now`? 

@dpallotti and @beckerhe might know more about this.

Reply:
Yeah, one is monotonic time since boot and the other is wall clock (the one we print in the capture log with "Capture started at ..."). Naming might be improvable but we want both.